### PR TITLE
Implemented decorators feature.

### PR DESCRIPTION
**Pull request by prowwid**
Basic implementation of[\[Feature Proposal\] Add Decorators](https://github.com/yiisoft/di/issues/10).

This PR contains a basic implementation of decorators to show how suggested decorators would look like. Examples of decorators present in tests. 

| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes

### CODE REVIEW

**Comment by prowwid** [comment]
Hello @samdark

Any new thoughts on this one?  
Maybe I need to add something?

**Comment by samdark** [comment]
Technically it is alright but I feel the same could be achieved in a simpler way.

Need opinions of @yiisoft/core-developers and @yiisoft/reviewers.

File: src/Container.php [hunk 1]
```diff
@@ -417,4 +428,58 @@ protected function buildProvider($providerDefinition)
...
+    public function addDecorator($id, $decorator)
+    {
+        if ($this->decoratorHasValidType($decorator)) {
+            $this->decorators[$id][] = $decorator;
+        } else {
+            throw new InvalidConfigException('Decorator should be a callable or implement ' . Decorator::class);
+        }
+    }
+
+    protected function decoratorHasValidType($decorator)
```
Comment (SilverFire) [inline src/Container.php]: How about `isDecorator($object)`?

File: src/Container.php [hunk 2]
```diff
@@ -417,4 +428,58 @@ protected function buildProvider($providerDefinition)
...
+        if (is_callable($decorator)) {
+            $isValid = true;
+        } elseif (is_object($decorator) && $decorator instanceof Decorator) {
+            $isValid = true;
+        } elseif (is_string($decorator) && class_exists($decorator) && (new ReflectionClass($decorator))->implementsInterface(Decorator::class)) {
+            $isValid = true;
+        } else {
+            $isValid = false;
+        }
+        return $isValid;
```
Comment (SilverFire) [inline src/Container.php]: Does this look more readable for you?

```php
if (is_callable($decorator)) {
    return true;
}
if (is_object($decorator) && $decorator instanceof Decorator) {
    return true;
}
if (is_string($decorator) && class_exists($decorator) && (new ReflectionClass($decorator))->implementsInterface(Decorator::class)) {
    return true;
}

return false;
```

File: src/Container.php [hunk 3]
```diff
@@ -417,4 +428,58 @@ protected function buildProvider($providerDefinition)
 
         return $provider;
     }
+
+    public function addDecorator($id, $decorator)
```
Comment (SilverFire) [inline src/Container.php]: `$id` should be renamed to `$groupName` for consistency with `runDecoratorsOnObject()` method

Comment (samdark) [reply src/Container.php]: phpdoc should be added as well...

File: src/Container.php [hunk 4]
```diff
@@ -56,6 +57,14 @@ class Container implements ContainerInterface
      * delayed to register till their services would be requested
      */
     private $delayedProviders;
+    /**
+     * @var contracts\Decorator[]|array
```
Comment (samdark) [inline src/Container.php]: Description should be added.

File: src/Container.php (hunk 1 above)
Comment (samdark) [reply src/Container.php]: phpdoc should be added.

File: src/Container.php [hunk 5]
```diff
@@ -417,4 +428,58 @@ protected function buildProvider($providerDefinition)
...
+            $isValid = true;
+        } elseif (is_string($decorator) && class_exists($decorator) && (new ReflectionClass($decorator))->implementsInterface(Decorator::class)) {
+            $isValid = true;
+        } else {
+            $isValid = false;
+        }
+        return $isValid;
+    }
+
+    protected function runDecoratorsOnObject($decoratorsGroupName, $object)
```
Comment (samdark) [inline src/Container.php]: phpdoc should be added.

File: src/Container.php [hunk 6]
```diff
@@ -417,4 +428,58 @@ protected function buildProvider($providerDefinition)
...
+    {
+        if (!isset($this->decorators[$decoratorsGroupName]) || empty($this->decorators[$decoratorsGroupName])) {
```
Comment (samdark) [inline src/Container.php]: Why do we need `!isset` check here?

File: src/Container.php [hunk 7]
```diff
@@ -417,4 +428,58 @@ protected function buildProvider($providerDefinition)
...
+                    '__class' => $decorator,
+                ]);
+
+                $this->decorators[$index] = $decorator;
+            }
+            $decorator->decorate($object);
+        }
+    }
+
+    protected function getInjector(): Injector
```
Comment (samdark) [inline src/Container.php]: phpdoc should be added.

File: src/contracts/Decorator.php [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ * Represents decorator of objects fetched from the container.
+ *
+ * @see https://sourcemaking.com/design_patterns/decorator
+ *
+ * @author Dmitry Kolodko <prowwid@gmail.com>
+ * @since 1.0
+ */
+interface Decorator
+{
+    public function decorate($object): void;
```
Comment (samdark) [inline src/contracts/Decorator.php]: phpdoc should be added.

**Comment by prowwid** [comment]
Hello @samdark and @SilverFire 

it was a draft realization for [\[Feature Proposal\] Add Decorators](https://github.com/yiisoft/di/issues/10) to reduce the talk on the issue so I basically didn't want to spend more time than needed in case we won't agree merging the PR.

I would add more docs later.

With best regards,
Dmitry

**Comment by samdark** [comment]
> I would add more docs later

**Comment by prowwid** [comment]
Hello @samdark and @SilverFire 

I've made changes in code based on your requests and added docs. Please review.

With best regards,
Dmitry

**Comment by samdark** [comment]
Made some adjustments.

**Comment by SilverFire** [comment]
Wait wait, we are definitely doing something wrong. The thing we've got makes possible to execute something on object right after its creation. Let's refer to the [Decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern) article, and read the following:

> Decorator objects implement the interface of the extended (decorated) object transparently by forwarding all requests to it and perform additional functionality before/after forwarding a request.

So current implementation is not a decorator. It should be called `post-configurator` instead, and it's definitely not what we need to do. This feature is absolutely useless, if the decorated object is immutable and does not have neither setters nor public properties.

There are a lot examples on different programming languages on Wikipedia, but all the decoration examples are about wrapping objects to provide additional features, not about changing decorated object internals.

**Comment by samdark** [comment]
Right. It was same concern on my part in https://github.com/yiisoft/di/issues/10#issuecomment-357382054 and I somehow missed when it went wrong :(

**Comment by samdark** [comment]
Anyway, while implementation is good and idea to have decorators is fine, I don't think we'll merge the pull request.

**Comment by SilverFire** [comment]
What we need is an API to build the following imperative construction in a declarative way:

```php
$car = new RemoteControlledCar(new ElectroCar(new Car()));
```

**Pull request closed without merging**